src/Modules/UngradedModules/FreeModuleHom.jl

```diff
@@ -450,40 +450,45 @@ represented as subquotient with no relations -> F)
 
 ```
 """
-function kernel(h::FreeModuleHom{<:FreeMod, <:FreeMod})  #ONLY for free modules...
-  error("not implemented for modules over rings of type $(typeof(base_ring(domain(h))))")
-end
-
-# The following function is part of the requirement of atomic functions to be implemented 
-# in order to have the modules run over a specific type of ring. The documentation of this is 
-# pending and so far only orally communicated by Janko Boehm. 
-#
-# The concrete method below uses Singular as a backend to achieve its task. In order 
-# to have only input which Singular can actually digest, we restrict the signature 
-# to those cases. The method used to be triggered eventually also for rings which 
-# did not have a groebner basis backend in Singular, but Singular did not complain. 
-# This lead to false results without notification. By restricting the signature, 
-# the user gets the above error message instead. 
-function kernel(
-    h::FreeModuleHom{<:FreeMod{T}, <:FreeMod{T}, Nothing}
-  ) where {S<: Union{ZZRingElem, <:FieldElem}, T <: MPolyRingElem{S}}
+function kernel(h::FreeModuleHom{<:FreeMod, <:FreeMod})
   is_zero(h) && return sub(domain(h), gens(domain(h)))
   is_graded(h) && return _graded_kernel(h)
-  return _simple_kernel(h)
+  return kernel_atomic(h)  # explicitly call kernel_atomic
+end
+
+function kernel_atomic(h::FreeModuleHom{<:FreeMod, <:FreeMod})
+  error("not implemented for modules over rings of type $(typeof(base_ring(domain(h))))")
 end
 
-function _simple_kernel(h::FreeModuleHom{<:FreeMod, <:FreeMod})
+function kernel_atomic(h::FreeModuleHom{<:FreeMod{T}, <:FreeMod{T}, Nothing}) where {S<:Union{ZZRingElem, <:FieldElem}, T<:MPolyRingElem{S}}
   F = domain(h)
   G = codomain(h)
-  g = images_of_generators(h)
-  b = ModuleGens(g, G, default_ordering(G))
-  M = syzygy_module(b)
+  gens_h = images_of_generators(h)
+  mod_gens = ModuleGens(gens_h, G, default_ordering(G))
+  M = syzygy_module(mod_gens)
   v = elem_type(F)[F(coordinates(repres(w))) for w in gens(M) if !is_zero(w)]
   return sub(F, v)
 end
 
+function ensure_kernel_ctx!(h::FreeModuleHom{<:FreeMod{T}, <:FreeMod{T}, Nothing}) where {T<:Union{ZZRingElem, FieldElem}}
+    if !has_attribute(h, :kernel_ctx)
+        mat_h = transpose(matrix(h))
+        set_attribute!(h, :kernel_ctx, solve_init(mat_h))
+    end
+end
+
+function kernel_atomic(h::FreeModuleHom{<:FreeMod{T}, <:FreeMod{T}, Nothing}) where {T<:Union{ZZRingElem, FieldElem}}
+    ensure_kernel_ctx!(h)
+    kernel_ctx = get_attribute(h, :kernel_ctx)
+    K = kernel(kernel_ctx, side=:right)
```

Comment:
this looks like it introduces something very similar to the `*_assure!` functions we are trying to get rid of.
Would it be possible to instead of calling `ensure_*` and then extracting an attribute just call a getter (that then makes use of `@attr` for caching)
```suggestion
function kernel_ctx(h::FreeModuleHom{<:FreeMod{T}, <:FreeMod{T}, Nothing}) where {T<:Union{ZZRingElem, FieldElem}}
    mat_h = transpose(matrix(h))
    return solve_init(mat_h)
end

function kernel_atomic(h::FreeModuleHom{<:FreeMod{T}, <:FreeMod{T}, Nothing}) where {T<:Union{ZZRingElem, FieldElem}}
    ctx = kernel_ctx(h)
    K = kernel(ctx; side=:right)
```